### PR TITLE
[inference] add smoothquant llama

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
- add smoothquant llama 
- smoothquant context attention and token attention
- example of smoothquant llama
- base smoothquant model 

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_smoothquant/test_llama_attention.py [hunk 1]
```diff
@@ -55,22 +54,34 @@ def torch_context_attention(xq, xk, xv, bs, seqlen, num_head, head_dim):
...
+    ones = torch.ones(hidden_size, hidden_size, dtype=torch.float, device="cuda")
 
     smooth_attn = smooth_attn.to("cuda")
 
-    input = torch.randint(-127, 127, (1, seq_len, head_num * head_dim), dtype=torch.int8, device="cuda")
+    input = torch.randint(-20, 20, (1, seq_len, head_num * head_dim), dtype=torch.int8, device="cuda")
+    input_scale = 1 / 20.0
+
+    output = torch.matmul(input.to(torch.float) * input_scale, ones)
+    qkv_max_out = torch.max(torch.abs(output)) / 127
```
Comment (yuanheng-zhao) [inline tests/test_smoothquant/test_llama_attention.py:81]: Just wonder: why 127

File: examples/inference/smoothquant_conversion.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    model = SmoothLlamaForCausalLM.from_quantized(output_path, model_basename="llama-7b")
+
+    generate_kwargs = dict(max_new_tokens=16, do_sample=False, use_cache=True)
+    input_tokens = tokenizer(["today is "], return_tensors="pt").to("cuda")
+    out = model.generate(**input_tokens, **generate_kwargs)
+    text = tokenizer.batch_decode(out)
+    print("out is:", text)
+
+
+if __name__ == "__main__":
```
Comment (yuanheng-zhao) [inline examples/inference/smoothquant_conversion.py:65]: Where are we going to use this script?

File: colossalai/inference/quant/smoothquant/models/llama.py [hunk 3]
```diff
@@ -225,7 +309,556 @@ def forward(
...
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (self_attn_weights,)
+
+        if use_cache:
+            outputs += (present_key_value,)
```
Comment (yuanheng-zhao) [inline colossalai/inference/quant/smoothquant/models/llama.py:433]: Do we really need this? I saw you already use the kv cache mechanism

File: tests/test_smoothquant/test_llama_attention.py (hunk 1 above)
Comment (Xu-Kai) [reply tests/test_smoothquant/test_llama_attention.py:81]: The range of int8 is [-128, 127]. When value divide the qkv_max_out, the ouput value can be in [-128, 127]

File: colossalai/inference/quant/smoothquant/models/llama.py (hunk 3 above)
Comment (Xu-Kai) [reply colossalai/inference/quant/smoothquant/models/llama.py]: Removed.

**Review by yuanheng-zhao — APPROVED** [review]

File: colossalai/inference/quant/smoothquant/models/llama.py (hunk 3 above)
Comment (yuanheng-zhao) [reply colossalai/inference/quant/smoothquant/models/llama.py]: Should also check if the removing of returning past_key_value would affect the flow in llama attention. (For original version it processes the passed-in past_key_value.)

Comment (Xu-Kai) [reply colossalai/inference/quant/smoothquant/models/llama.py]: I have checked and updated the flow.

**Pull request merged**